File: src/main/java/com/naya/spark/dataFrame/java/ProfileServiceSpark.java
package com.naya.spark.dataFrame.java;

import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import scala.Tuple2;
import java.util.Arrays;

import static org.apache.spark.sql.functions.*;

public class ProfileServiceSpark {

    public void printTable(Dataset<Row> dataset){
        dataset.show();
    }

    public void printSchema(Dataset<Row> dataset){
        dataset.printSchema();
    }

    public void printTypes(Dataset<Row> dataset){
        Tuple2<String, String>[] dtypes = dataset.dtypes();
        Arrays.stream(dtypes).forEach(System.out::println);
    }

    public Dataset<Row> addColSalary(Dataset<Row> dataset){
        return dataset.withColumn("salary", col("age").multiply(size(col("keywords")).multiply(10)));
    }

    public Dataset<Row> findProgrammersWithMostPopularTechnologyWithSalaryLower1200(Dataset<Row> dataset){
        Dataset<Row> persist = dataset.withColumn("tech", explode(col("keywords")))
                .persist();

        String mostPopular = persist.select(col("tech"))
                .groupBy("tech")
                .count()
                .sort(desc("count"))
                .first().get(0).toString();

        System.out.println(mostPopular);

        return persist.filter(col("tech").equalTo(mostPopular))
                .filter(col("salary").leq(1200))
                .select("name");
    }
}

File: src/main/java/com/naya/spark/dataFrame/scala/ProfileServiceSpark.scala
package com.naya.spark.dataFrame.scala

import org.apache.spark.sql.{DataFrame, RelationalGroupedDataset}
import org.apache.spark.sql.functions._

class ProfileServiceSpark {

  def printTable(dataFrame: DataFrame):Unit={
    dataFrame.show()
  }

  def printSchema(dataFrame: DataFrame):Unit= {
    dataFrame.printSchema()
  }

  def printTypes(dataFrame: DataFrame):Unit= {
    val dtypes = dataFrame.dtypes
    dtypes.foreach(println(_))
  }

  def addColSalary(dataFrame: DataFrame):DataFrame= {
    dataFrame.withColumn("salary", col("age") * size(col("keywords")) * 10)
  }

  def findProgrammersWithMostPopularTechnologyWithSalaryLower1200(dataFrame: DataFrame):DataFrame= {
    val dataFrame1 = dataFrame.withColumn("tech", explode(col("keywords")))
      .persist()

    val mostPopular = dataFrame1.select(col("tech"))
      .groupBy("tech").count.sort(desc("count"))
      .first.get(0)

    println(mostPopular)

    dataFrame1.filter(col("tech").equalTo(mostPopular))
      .filter(col("salary").leq(1200))
      .select("name")
  }
}
